src/Type/Nette/JsonDecodeDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,143 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Nette;
+
+use Nette\Utils\Json;
+use PhpParser\Node\Arg;
+use PhpParser\Node\Expr\ClassConstFetch;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\Identifier;
+use PhpParser\Node\Name;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\BooleanType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicStaticMethodReturnTypeExtension;
+use PHPStan\Type\FloatType;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\UnionType;
+use stdClass;
+
+final class JsonDecodeDynamicReturnTypeExtension implements DynamicStaticMethodReturnTypeExtension
+{
+
+	public function getClass(): string
+	{
+		return 'Nette\Utils\Json';
+	}
+
+	public function isStaticMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return $methodReflection->getName() === 'decode';
+	}
+
+	public function getTypeFromStaticMethodCall(MethodReflection $methodReflection, StaticCall $methodCall, Scope $scope): Type
+	{
+		$args = $methodCall->getArgs();
+
+		$isForceArray = $this->isForceArray($args);
+
+		$firstArgValue = $args[0]->value;
+		$firstValueType = $scope->getType($firstArgValue);
+
+		if ($firstValueType instanceof ConstantStringType) {
+			$resolvedType = $this->resolveConstantStringType($firstValueType, $isForceArray);
+		} else {
+			$resolvedType = new MixedType();
+		}
+
+		if (! $resolvedType instanceof MixedType) {
+			return $resolvedType;
+		}
+
+		// fallback type
+		if ($isForceArray) {
+			return new UnionType([
+				new ArrayType(new MixedType(), new MixedType()),
+				new StringType(),
+				new FloatType(),
+				new IntegerType(),
+				new BooleanType(),
+			]);
+		}
+
+		// scalar types with stdClass
+		return new UnionType([
+			new ObjectType(stdClass::class),
```

Comment:
It can still be an array here.

Reply:
On what input?

Reply:
I guess the stdclass could contain a array of other things?

Reply:
You don't have to guess: https://3v4l.org/qsjVh

BTW I worry that these precise types would become really annoying. Right now people working with json_decode() get some errors on level 9, but with these union types it would be much sonner - on level 7. I think the best course of action is to return "MixedType without stdClass" with FORCE_ARRAY, and normal "MixedType" without FORCE_ARRAY.

MixedType is subtractable - `new MixedType(true, new ObjectType(\stdClass::class))` means `mixed without stdClass`.

Reply:
> MixedType is subtractable

Cool, that's a new trick to me.

---

>  I think the best course of action is to return "MixedType without stdClass" with FORCE_ARRAY, and normal "MixedType" without FORCE_ARRAY.

Do you mean like this?

```php
if ($isForceArray) {
    return new MixedType(true, new ObjectType(\stdClass::class));
}

return new Mixed(true);
```

It might be change with more strict approach, but it would be more correct. The reason we made this extension for our project is to make sure there is only scalar value or stdClass. Mixed includes anything, objects, array of object etc.

Yet the DX is important, I understand. It might be useful to enable this only since level 9+.